### ISSUE TITLE
Create a better way to manage subproject docs

### ISSUE THREAD

**Issue by ellisonbg** [jupyter/jupyter#51]
@willingc has been thinking about how to manage the fact that the different Jupyter/IPython subprojects are in lots of repos. That presents a number of problems for users:

* When they go between each RTD repo, they home icon in RTD changes its target.
* Then TOC for the docs changes to that repo.
* There isn't any easy way to get back to the global index.
* There isn't an easy link to the jupyter.org or ipython.org home pages.

These things were made painfully clear in some of the users tests we ran.

I wanted to open an issue to use in discussing solutions to these problems.

Overall, I think we should start with the simplest thing possible, and test user test it to improve it. 

I think the simplest thing would be to create a single Subproject index page that all of our RTD repos can manually have a copy of. This would allow us to focus on the visual and information design of a single page and have that replicated across our repos, without having to get into combining all the docs into a single repo or using git submodules. 

Thoughts?

@minrk @takluyver @fperez

**Comment by minrk** [comment]
It's difficult. I think we should push as much docs as we can to jupyter/jupyter, but that has to exclude version-sensitive things, since versions of different projects are now independent of each other. At the very least, API docs have to be in their own repos, and probably most docs that include *calling* APIs need to, as well. We should definitely be linking heavily, and probably ensure that we get a link to jupyter/jupyter in the sidebar of all the docs.

**Comment by ellisonbg** [comment]
Min, are you thinking we should move existing docs in subrepos to
jupyter/jupyter? That would definitely simplify some of these issues - as
long as we *move* and don't duplicate content.

On Thu, Oct 8, 2015 at 8:15 AM, Min RK <notifications@github.com> wrote:

> It's difficult. I think we should push as much docs as we can to
> jupyter/jupyter, but that has to exclude version-sensitive things, since
> versions of different projects are now independent of each other. At the
> very least, API docs have to be in their own repos, and probably most docs
> that include *calling* APIs need to, as well. We should definitely be
> linking heavily, and probably ensure that we get a link to jupyter/jupyter
> in the sidebar of all the docs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/jupyter/issues/51#issuecomment-146519578>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by tacaswell** [comment]
We have a similar problem at BNL.

attn @danielballan @ericdill

**Comment by minrk** [comment]
@ellisonbg I'm not sure if/how much we should move. It's a hard line to draw, since we are *trying* to make most of these decoupled projects. I think jupyter/jupyter should probably be the place for most high level, non-version-sensitive docs, but that's a nebulous criterion and I'm not sure how useful it is. I do think we *will* have docs in several repos, and that means heavy amounts of inter-linking is going to be necessary, no matter how much of the documentation ends up at jupyter/jupyter, since it's always going to exclude any API docs (including hand-written, not just autogen-api). And to some degree, if we do a good job with linking/cross-repo navigation, then it becomes less important that a particular doc is in the 'right' repo.

A judicious pass on most pages of "Where am I likely to go next" should give a sense of what links need to be added. We also need to make sure that we treat rst anchors in our various docs as public API link targets, so we don't break intersphinx unnecessarily.

**Comment by willingc** [comment]
@minrk @ellisonbg May I have sufficient permissions to label issues and PRs on all jupyter repos (including subprojects)? I would like to set up a consistent documentation label for all repos (with a pastel color) so that triage and tracking of doc issues will be simplified for all developers.

As for overall docs approach, let's try to move forward in phases (that will enable positive forward progress yet not preclude more flexible decisions moving forward):

**Step 1 (Landing page)**
I think an excellent first step for consolidating the docs would be to make a simple landing page with visually appealing icons as @ellisonbg has mentioned before. Perhaps hosted as docs.jupyter.org. Each icon would be clickable and direct to the subproject's documentation (wherever it is currently hosted). 

**Step 2 (Redirect link back to landing page)**
The next phase would be adding a simple redirect link back to the `docs.jupyter.org` landing page in each subproject's table of contents.

**Step 3 (Work on interproject linking of docs)**
This is going to be a spaghetti step with whatever approach is taken. Given the priorities with triaging existing issues and PRs as well as current developer bandwidth, I think that it would be prudent (especially for our end users) to tackle step 1 and 2. This will give the overall effort structure and from there make it much easier to tackle step 3 with minimal disruption to the current features and future functionality that is accelerating in all the projects.

**Comment by tacaswell** [comment]
@willingc were you planning to do the labeling by hand or write some tooling to work with gh api?  

I have some hacked-code that sorta does the label normalization and for day job would definitely be interested in tooling to make that easy to manage.


-----
super hacked code:

```python
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)
import six

from github import Github

make_labels = {'pb.HXN': '5319e7',
               'pb.XPD': '5319e7',
               'pb.IXS':'5319e7',
               'pb.CHX': '5319e7',
               'pb.CSX': '5319e7',
               'Needs Review': '009800',
               'pb.SRX': '5319e7',
               'vistrails': 'fbca04',
               'Beamline Commissioning': 'fad8c7',
               'Early Science': 'fad8c7',
               'Needs Revision': 'e11d21'}

with open(os.path.expanduser('~/.ghoauth')) as fin:
    token = fin.readline()

g = Github(token)
org = g.get_organization('NSLS-II')
repo_list  = ['vistools',
              'userpackages',
              'KitchenSink',
              'Internal',
              'Controls',
              'NSLS2']


for r_name in repo_list:
    repo = org.get_repo(r_name)
    current_labels = set(l.name for l in repo.get_labels())

    for name, color in six.iteritems(make_labels):
        # if we don't have this label, make it
        if name not in current_labels:
            repo.create_label(name, color)
```

**Comment by willingc** [comment]
@tacaswell I was planning to do the initial tagging by hand. Primarily because I wish to read each issue.

That said, thank you for sharing the tool, and I agree automating down the road would be great. Have you seen/used github3py https://jupyter-notebook.readthedocs.org/en/latest/#developer-documentation ? The maintainer @sigmavirus24 is a fabulous developer.

**Comment by tacaswell** [comment]
No, when I wrote the above I spent a while jumping between python wrappings of the gh api before I found one that worked 'well enough'.  I am mostly thinking about the mechanical task of creating/re-coloring all of the labels, not applying them to the individual issues.  Something like `ensure_repo_labels(repo, label_color_dict, token)`.

**Comment by willingc** [comment]
@tacaswell This is some old code of a project when we were triaging OpenHatch issues on GH after migrating them from Bugzilla. I found using the Jupyter notebook as well as Pandas helpful for sanity checking any automation. https://github.com/willingc/openhatch-issues

**Comment by minrk** [comment]
@willingc do you have access to create and/or assign labels on this repo? I'm trying to figure out the permissions bit.

**Comment by willingc** [comment]
@minrk Nope. Thanks for the IPython access for labels.

**Comment by sigmavirus24** [comment]
I'm happy to help with API scripting as needed but it looks like @tacaswell has figured something out. I'm going to unsubscribe from the issue but @willingc knows where to find me on IRC if needed.

**Comment by willingc** [comment]
@sigmavirus24 Thanks. Tagged you early just because I like your project. :surfer:
